src/services/horoscope-handler/handler.service.ts

```diff
@@ -0,0 +1,96 @@
+import {
+  Action,
+  Service,
+} from '@ourparentcenter/moleculer-decorators-extended';
+import { Context, ServiceBroker } from 'moleculer';
+import BaseService from '../../base/base.service';
+import { SERVICE } from '../../common';
+import { Block, Transaction } from '../../models';
+import networks from '../../../network.json' assert { type: 'json' };
+
+@Service({
+  name: SERVICE.V1.HoroscopeHandlerService.key,
+  version: 1,
+})
+export default class HoroscopeHandlerService extends BaseService {
+  public constructor(public broker: ServiceBroker) {
+    super(broker);
+  }
+
+  @Action({
+    name: SERVICE.V1.HoroscopeHandlerService.getData.key,
+    params: {
+      chainid: {
+        type: 'string',
+        optional: false,
+        enum: networks.map((network) => network.chainId),
+      },
+      startBlock: {
+        type: 'number',
+        optional: false,
+      },
+      endBlock: {
+        type: 'number',
+        optional: false,
+      },
+    },
+  })
+  public async getData(
+    ctx: Context<
+      { chainid: string; startBlock: number; endBlock: number },
+      Record<string, unknown>
+    >
+  ) {
+    // TODO: handler filter from request
+
+    // query to get data
+    const { startBlock, endBlock } = ctx.params;
+    const queryBlock = Block.query()
+      .select('height', 'hash', 'time')
+      .withGraphFetched('events.[attributes]')
+      .modifyGraph('events', (builder) => {
+        builder.select('type', 'source');
+      })
+      .modifyGraph('events.[attributes]', (builder) => {
+        builder.select('key', 'value');
+      })
+      .where('height', '>=', startBlock)
+      .andWhere('height', '<', endBlock);
+
+    const queryTransaction = Transaction.query()
+      .select('height', 'hash', 'code', 'codespace', 'memo', 'index')
+      .withGraphFetched('messages')
+      .withGraphFetched('events.[attributes]')
+      .modifyGraph('messages', (builder) => {
+        builder
+          .select('type', 'sender', 'content', 'index')
+          .whereNull('parent_id');
+      })
+      .modifyGraph('events', (builder) => {
+        builder.select('type', 'source');
+      })
+      .modifyGraph('events.[attributes]', (builder) => {
+        builder.select('key', 'value');
+      })
+      .where('height', '>=', startBlock)
+      .andWhere('height', '<', endBlock);
+    const [resultBlock, resultTransaction] = await Promise.all([
+      queryBlock,
+      queryTransaction,
+    ]);
+    resultTransaction.forEach((transaction) => {
```

Comment:
chỗ này thêm order height cho cả block với tx rồi cứ thế đẩy vào thôi

Reply:
mà sao ko join luôn được ở trên em? nó to quá à?

Reply:
em tách ra để handle filter ở từng câu cho dễ ấy

Reply:
ok

Reply:
lúc sau có filter em phải thêm 1 đống where vào, sợ để ở 1 câu thì to quá nên tách ra

Reply:
thế thêm order rồi sửa cái vòng for là được

Reply:
sửa vòng for là sao anh nhỉ

Reply:
fixed